File: tools/harness/artifacts.ts
import { randomBytes } from "node:crypto";
import { mkdir, readFile, readdir, rename, stat, writeFile } from "node:fs/promises";
import { basename, dirname, join, relative, sep } from "node:path";

import { canonicalJsonBytes, sha256Hex } from "@palimpsest/contracts";

import { attemptPath, type HarnessAttemptIdentity } from "./config.js";

export type AttemptClassification = "completed" | "failed" | "invalid";
export type AttemptFailurePhase = "run" | "grade" | "replay" | "complete";

interface AttemptReceipt {
  schemaVersion: 1;
  attemptId: string;
  declarationDigest: string;
  runId: string;
  startedAt: string;
  phase: "running";
}

interface OutputEntry {
  path: string;
  byteLength: number;
  sha256: string;
}

export interface TerminalAttempt {
  schemaVersion: 1;
  attemptId: string;
  declarationDigest: string;
  runId: string;
  startedAt: string;
  classification: AttemptClassification;
  outputs: OutputEntry[];
}

interface AttemptFailure {
  schemaVersion: 1;
  declarationDigest: string;
  runId: string;
  phase: AttemptFailurePhase;
  errorName: string;
  message: string;
}

function attemptId(identity: HarnessAttemptIdentity): string {
  return `harness/${identity.declarationDigest}/${identity.runId}`;
}

async function pathExists(path: string): Promise<boolean> {
  try {
    await stat(path);
    return true;
  } catch (error) {
    if (error instanceof Error && "code" in error && error.code === "ENOENT") {
      return false;
    }
    throw error;
  }
}

async function atomicJson(path: string, value: unknown): Promise<void> {
  const temporary = join(
    dirname(path),
    `.${basename(path)}.${process.pid}.${randomBytes(4).toString("hex")}.tmp`,
  );
  await writeFile(temporary, canonicalJsonBytes(value), { flag: "wx" });
  await rename(temporary, path);
}

async function outputPaths(root: string, current = root): Promise<string[]> {
  const paths: string[] = [];
  for (const name of (await readdir(current)).sort()) {
    if (current === root && name === "terminal.json") {
      continue;
    }
    const path = join(current, name);
    const metadata = await stat(path);
    if (metadata.isDirectory()) {
      paths.push(...(await outputPaths(root, path)));
    } else if (metadata.isFile()) {
      paths.push(relative(root, path).split(sep).join("/"));
    } else {
      throw new Error(`Attempt output is not a regular file or directory: ${path}`);
    }
  }
  return paths.sort();
}

async function outputEntries(root: string): Promise<OutputEntry[]> {
  return Promise.all(
    (await outputPaths(root)).map(async (path) => {
      const bytes = await readFile(join(root, path));
      return {
        path,
        byteLength: bytes.byteLength,
        sha256: sha256Hex(bytes),
      };
    }),
  );
}

async function writeCurrent(
  root: string,
  identity: HarnessAttemptIdentity,
  path: string,
  status: "running" | AttemptClassification,
): Promise<void> {
  await atomicJson(join(root, "current.json"), {
    schemaVersion: 1,
    attemptId: attemptId(identity),
    declarationDigest: identity.declarationDigest,
    runId: identity.runId,
    attemptPath: path,
    status,
    evidence: false,
  });
}

export async function createAttempt(options: {
  root: string;
  identity: HarnessAttemptIdentity;
  startedAt: string;
}): Promise<string> {
  const { root, identity, startedAt } = options;
  const path = attemptPath(root, identity);
  await mkdir(dirname(path), { recursive: true });
  await mkdir(path, { recursive: false });
  const receipt: AttemptReceipt = {
    schemaVersion: 1,
    attemptId: attemptId(identity),
    declarationDigest: identity.declarationDigest,
    runId: identity.runId,
    startedAt,
    phase: "running",
  };
  await writeFile(join(path, "attempt.json"), canonicalJsonBytes(receipt), { flag: "wx" });
  await writeCurrent(root, identity, path, "running");
  return path;
}

export async function sealAttempt(options: {
  root: string;
  identity: HarnessAttemptIdentity;
  classification: AttemptClassification;
}): Promise<TerminalAttempt> {
  const { root, identity, classification } = options;
  const path = attemptPath(root, identity);
  const terminalPath = join(path, "terminal.json");
  if (await pathExists(terminalPath)) {
    throw new Error(`Attempt already has a terminal manifest: ${terminalPath}`);
  }
  const receipt = JSON.parse(await readFile(join(path, "attempt.json"), "utf8")) as AttemptReceipt;
  if (
    receipt.attemptId !== attemptId(identity) ||
    receipt.declarationDigest !== identity.declarationDigest ||
    receipt.runId !== identity.runId
  ) {
    throw new Error("Attempt receipt identity does not match the explicit attempt.");
  }
  const terminal: TerminalAttempt = {
    schemaVersion: 1,
    attemptId: receipt.attemptId,
    declarationDigest: receipt.declarationDigest,
    runId: receipt.runId,
    startedAt: receipt.startedAt,
    classification,
    outputs: await outputEntries(path),
  };
  await atomicJson(terminalPath, terminal);
  await writeCurrent(root, identity, path, classification);
  return terminal;
}

function failureRecord(
  identity: HarnessAttemptIdentity,
  phase: AttemptFailurePhase,
  error: unknown,
): AttemptFailure {
  return {
    schemaVersion: 1,
    declarationDigest: identity.declarationDigest,
    runId: identity.runId,
    phase,
    errorName: error instanceof Error ? error.name : "NonErrorFailure",
    message: error instanceof Error ? error.message : String(error),
  };
}

export async function sealFailedAttempt(options: {
  root: string;
  identity: HarnessAttemptIdentity;
  phase: AttemptFailurePhase;
  error: unknown;
}): Promise<TerminalAttempt> {
  const { root, identity, phase, error } = options;
  const path = attemptPath(root, identity);
  const terminalPath = join(path, "terminal.json");
  if (await pathExists(terminalPath)) {
    const terminal = await verifyTerminalAttempt({ root, identity });
    await writeCurrent(root, identity, path, terminal.classification);
    return terminal;
  }
  const failurePath = join(path, "failure.json");
  const failure = failureRecord(identity, phase, error);
  if (await pathExists(failurePath)) {
    const prior = JSON.parse(await readFile(failurePath, "utf8")) as unknown;
    if (!canonicalJsonBytes(prior).equals(canonicalJsonBytes(failure))) {
      throw new Error("Attempt failure evidence conflicts with the existing failure record.");
    }
  } else {
    await atomicJson(failurePath, failure);
  }
  return sealAttempt({ root, identity, classification: "failed" });
}

export async function sealCurrentAttemptFailure(options: {
  root: string;
  runId: string;
  phase: AttemptFailurePhase;
  error: unknown;
}): Promise<TerminalAttempt | null> {
  const { root, runId, phase, error } = options;
  const currentPath = join(root, "current.json");
  if (!(await pathExists(currentPath))) return null;
  const current = JSON.parse(await readFile(currentPath, "utf8")) as Record<string, unknown>;
  if (
    current.runId !== runId ||
    current.status !== "running" ||
    typeof current.declarationDigest !== "string" ||
    typeof current.attemptPath !== "string"
  ) {
    return null;
  }
  const identity = { declarationDigest: current.declarationDigest, runId };
  if (current.attemptPath !== attemptPath(root, identity)) {
    throw new Error("Current attempt pointer does not match its declaration-bound path.");
  }
  return sealFailedAttempt({
    root,
    identity,
    phase,
    error,
  });
}

export async function verifyTerminalAttempt(options: {
  root: string;
  identity: HarnessAttemptIdentity;
}): Promise<TerminalAttempt> {
  const { root, identity } = options;
  const path = attemptPath(root, identity);
  const terminal = JSON.parse(
    await readFile(join(path, "terminal.json"), "utf8"),
  ) as TerminalAttempt;
  if (
    terminal.attemptId !== attemptId(identity) ||
    terminal.declarationDigest !== identity.declarationDigest ||
    terminal.runId !== identity.runId
  ) {
    throw new Error("Terminal manifest identity does not match the explicit attempt.");
  }
  const actual = await outputEntries(path);
  if (!canonicalJsonBytes(actual).equals(canonicalJsonBytes(terminal.outputs))) {
    throw new Error("Attempt directory does not match the terminal exact output set.");
  }
  return terminal;
}

File: python/src/palimpsest/gate_c/artifacts.py
from __future__ import annotations

import json
import os
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from palimpsest.contracts import canonical_json_bytes, sha256_hex
from palimpsest.gate_b.artifacts import write_canonical as _write_canonical
from palimpsest.gate_b.artifacts import write_text as _write_text

SHA256_PATTERN = re.compile(r"^[0-9a-f]{64}$")
RUN_ID_PATTERN = re.compile(r"^[a-z0-9][a-z0-9-]{0,63}$")


@dataclass(frozen=True)
class AttemptIdentity:
    declaration_digest: str
    run_id: str

    def __post_init__(self) -> None:
        if not SHA256_PATTERN.fullmatch(self.declaration_digest):
            raise ValueError("Declaration digest must be 64 lowercase hexadecimal characters.")
        if not RUN_ID_PATTERN.fullmatch(self.run_id):
            raise ValueError("Run ID must be a lowercase slug.")

    @property
    def attempt_id(self) -> str:
        return f"gate-c/{self.declaration_digest}/{self.run_id}"


def write_canonical(path: Path, value: Any) -> dict[str, Any]:
    return _write_canonical(path, value)


def write_text(path: Path, value: str, artifact_type: str) -> dict[str, Any]:
    return _write_text(path, value, artifact_type)


def attempt_path(attempts_root: Path, identity: AttemptIdentity) -> Path:
    return attempts_root / identity.declaration_digest / identity.run_id


def _atomic_write(path: Path, content: bytes) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    temporary = path.parent / f".{path.name}.{os.getpid()}.tmp"
    temporary.write_bytes(content)
    os.replace(temporary, path)


def write_current_pointer(
    current_path: Path,
    *,
    identity: AttemptIdentity,
    path: Path,
    started_at: str,
    status: str,
) -> None:
    pointer = {
        "schemaVersion": 1,
        "attemptId": identity.attempt_id,
        "declarationDigest": identity.declaration_digest,
        "runId": identity.run_id,
        "attemptPath": path.as_posix(),
        "startedAt": started_at,
        "status": status,
        "evidence": False,
    }
    _atomic_write(current_path, canonical_json_bytes(pointer))


def create_attempt(
    *,
    attempts_root: Path,
    current_path: Path,
    identity: AttemptIdentity,
    started_at: str,
) -> Path:
    path = attempt_path(attempts_root, identity)
    path.mkdir(parents=True, exist_ok=False)
    manifest = {
        "schemaVersion": 1,
        "attemptId": identity.attempt_id,
        "declarationDigest": identity.declaration_digest,
        "runId": identity.run_id,
        "startedAt": started_at,
        "phase": "running",
    }
    write_canonical(path / "attempt.json", manifest)
    write_current_pointer(
        current_path,
        identity=identity,
        path=path,
        started_at=started_at,
        status="running",
    )
    return path


def resolve_attempt(
    *,
    attempts_root: Path,
    identity: AttemptIdentity,
) -> Path:
    path = attempt_path(attempts_root, identity)
    manifest_path = path / "attempt.json"
    if not manifest_path.is_file():
        raise FileNotFoundError(f"Attempt manifest does not exist: {manifest_path}")
    manifest = json.loads(manifest_path.read_text(encoding="utf-8"))
    expected = {
        "attemptId": identity.attempt_id,
        "declarationDigest": identity.declaration_digest,
        "runId": identity.run_id,
    }
    for field, value in expected.items():
        if manifest.get(field) != value:
            raise ValueError(f"Attempt manifest {field} does not match the explicit identity.")
    return path


def _output_entries(path: Path) -> list[dict[str, Any]]:
    entries: list[dict[str, Any]] = []
    outputs = [
        candidate
        for candidate in path.rglob("*")
        if candidate.is_file() and candidate.name != "terminal.json"
    ]
    for output in sorted(outputs, key=lambda candidate: candidate.relative_to(path).as_posix()):
        content = output.read_bytes()
        entries.append(
            {
                "path": output.relative_to(path).as_posix(),
                "byteLength": len(content),
                "sha256": sha256_hex(content),
            }
        )
    return entries


def finalize_attempt(
    *,
    path: Path,
    identity: AttemptIdentity,
    status: str,
    terminal_fields: dict[str, Any],
) -> dict[str, Any]:
    terminal_path = path / "terminal.json"
    if terminal_path.exists():
        raise FileExistsError(f"Attempt is already terminal: {terminal_path}")
    manifest = json.loads((path / "attempt.json").read_text(encoding="utf-8"))
    terminal = {
        "schemaVersion": 1,
        "attemptId": identity.attempt_id,
        "declarationDigest": identity.declaration_digest,
        "runId": identity.run_id,
        "status": status,
        **terminal_fields,
        "outputs": _output_entries(path),
    }
    if manifest.get("attemptId") != identity.attempt_id:
        raise ValueError("Attempt start manifest does not match terminal identity.")
    _atomic_write(terminal_path, canonical_json_bytes(terminal))
    return terminal


def resolve_terminal_attempt(
    *,
    attempts_root: Path,
    identity: AttemptIdentity,
) -> tuple[Path, dict[str, Any]]:
    path = resolve_attempt(attempts_root=attempts_root, identity=identity)
    terminal_path = path / "terminal.json"
    if not terminal_path.is_file():
        raise FileNotFoundError(f"Attempt has no terminal manifest: {terminal_path}")
    terminal = json.loads(terminal_path.read_text(encoding="utf-8"))
    for field, value in {
        "attemptId": identity.attempt_id,
        "declarationDigest": identity.declaration_digest,
        "runId": identity.run_id,
    }.items():
        if terminal.get(field) != value:
            raise ValueError(f"Terminal manifest {field} does not match explicit identity.")
    if terminal.get("outputs") != _output_entries(path):
        raise ValueError("Terminal manifest does not match the exact attempt output set.")
    return path, terminal
